pkg/controller/node/ipam/cloud_cidr_allocator.go

```diff
@@ -42,84 +45,149 @@ import (
+func (ca *cloudCIDRAllocator) removeNodeFromProcessing(nodeName string) {
+	ca.Lock()
+	defer ca.Unlock()
+	ca.nodesInProcessing.Delete(nodeName)
+}
 
+// WARNING: If you're adding any return calls or defer any more work from this
+// function you have to handle correctly nodesInProcessing.
+func (ca *cloudCIDRAllocator) AllocateOrOccupyCIDR(node *v1.Node) error {
+	if node == nil {
+		return nil
+	}
+	if !ca.insertNodeToProcessing(node.Name) {
+		glog.V(2).Infof("Node %v is already in a process of CIDR assignment.", node.Name)
+		return nil
+	}
 	cidrs, err := ca.cloud.AliasRanges(types.NodeName(node.Name))
-
 	if err != nil {
+		ca.removeNodeFromProcessing(node.Name)
 		util.RecordNodeStatusChange(ca.recorder, node, "CIDRNotAvailable")
 		return fmt.Errorf("failed to allocate cidr: %v", err)
 	}
-
 	if len(cidrs) == 0 {
+		ca.removeNodeFromProcessing(node.Name)
 		util.RecordNodeStatusChange(ca.recorder, node, "CIDRNotAvailable")
-		glog.V(2).Infof("Node %v has no CIDRs", node.Name)
-		return fmt.Errorf("failed to allocate cidr (none exist)")
+		return fmt.Errorf("failed to allocate cidr: Node %v has no CIDRs", node.Name)
 	}
 
-	node, err = ca.client.Core().Nodes().Get(node.Name, metav1.GetOptions{})
-	if err != nil {
-		glog.Errorf("Could not get Node object from Kubernetes: %v", err)
-		return err
+	glog.V(10).Infof("Putting node %s with CIDR %s into the work queue", node.Name, cidrs[0])
+	ca.nodeCIDRUpdateChannel <- nodeAndCIDR{
+		nodeName: node.Name,
+		cidr:     cidrs[0],
 	}
+	return nil
+}
 
-	podCIDR := cidrs[0]
-
-	if node.Spec.PodCIDR != "" {
-		if node.Spec.PodCIDR == podCIDR {
-			glog.V(3).Infof("Node %v has PodCIDR %v", node.Name, podCIDR)
-			return nil
+// Assigns CIDR to Node and sends an update to the API server.
+func (ca *cloudCIDRAllocator) updateCIDRAllocation(data nodeAndCIDR) error {
+	var err error
+	var node *v1.Node
+	defer ca.removeNodeFromProcessing(data.nodeName)
+	podCIDR := data.cidr
+	for rep := 0; rep < cidrUpdateRetries; rep++ {
+		// TODO: change it to using PATCH instead of full Node updates.
+		node, err = ca.client.Core().Nodes().Get(data.nodeName, metav1.GetOptions{})
+		if err != nil {
+			glog.Errorf("Failed while getting node %v to retry updating Node.Spec.PodCIDR: %v", data.nodeName, err)
+			continue
+		}
+		if node.Spec.PodCIDR != "" {
+			if node.Spec.PodCIDR == podCIDR {
+				glog.V(4).Infof("Node %v already has allocated CIDR %v. It matches the proposed one.", node.Name, podCIDR)
+				return nil
+			}
+			glog.Errorf("PodCIDR cannot be reassigned, node %v spec has %v, but cloud provider has assigned %v",
+				node.Name, node.Spec.PodCIDR, podCIDR)
+			// We fall through and set the CIDR despite this error. This
+			// implements the same logic as implemented in the
+			// rangeAllocator.
+			//
+			// See https://github.com/kubernetes/kubernetes/pull/42147#discussion_r103357248
+		}
+		node.Spec.PodCIDR = podCIDR
+		if _, err = ca.client.Core().Nodes().Update(node); err != nil {
```

Comment:
seems better to handle 

err == nil case in the `if` statement and return

```go
if _, err = ca.client.Core().Nodes().Update(node); err == nil {
  glog.Infof("Set node %q PodCIDR to %v", ...)
  // emit an event
  return nil
}

glog.Errorf("Failed to update node %q PodCIDR to %v (attempt %d): %v", node.Name, podCIDR, rep+1, err)
```

Reply:
+1

Reply:
Done